File: src/common/JsonNode.cs
using System.Text.Json;
using System;
using System.Text.Json.Nodes;

namespace common;

public static class JsonNodeModule
{
    public static JsonResult<JsonObject> AsJsonObject(this JsonNode? node) =>
        node switch
        {
            JsonObject jsonObject => JsonResult.Succeed(jsonObject),
            null => JsonResult.Fail<JsonObject>("JSON node is null."),
            _ => JsonResult.Fail<JsonObject>("JSON node is not a JSON object.")
        };

    public static JsonResult<JsonArray> AsJsonArray(this JsonNode? node) =>
        node switch
        {
            JsonArray jsonArray => JsonResult.Succeed(jsonArray),
            null => JsonResult.Fail<JsonArray>("JSON node is null."),
            _ => JsonResult.Fail<JsonArray>("JSON node is not a JSON array.")
        };

    public static JsonResult<JsonValue> AsJsonValue(this JsonNode? node) =>
        node switch
        {
            JsonValue jsonValue => JsonResult.Succeed(jsonValue),
            null => JsonResult.Fail<JsonValue>("JSON node is null."),
            _ => JsonResult.Fail<JsonValue>("JSON node is not a JSON value.")
        };

    public static JsonResult<T> Deserialize<T>(BinaryData? data, JsonSerializerOptions? options = default)
    {
        if (data is null)
        {
            return JsonResult.Fail<T>("Binary data is null.");
        }

        try
        {
            var jsonObject = JsonSerializer.Deserialize<T>(data, options ?? JsonSerializerOptions.Web);

            return jsonObject is null
                ? JsonResult.Fail<T>("Deserialization return a null result.")
                : JsonResult.Succeed(jsonObject);
        }
        catch (JsonException exception)
        {
            var jsonError = JsonError.From(exception);
            return JsonResult.Fail<T>(jsonError);
        }
    }
}

File: src/fsharp/JsonNode.fs
[<RequireQualifiedAccess>]
module common.JsonNode

open System
open System.Text.Json.Nodes
open System.IO

let asJsonObject (node: JsonNode | null) =
    match node with
    | :? JsonObject as jsonObject -> JsonResult.succeed jsonObject
    | Null -> JsonResult.failWithMessage "JSON node is null"
    | _ -> JsonResult.failWithMessage "JSON node is not a JSON object"

let asJsonArray (node: JsonNode | null) =
    match node with
    | :? JsonArray as jsonArray -> JsonResult.succeed jsonArray
    | Null -> JsonResult.failWithMessage "JSON node is null"
    | _ -> JsonResult.failWithMessage "JSON node is not a JSON array"

let asJsonValue (node: JsonNode | null) =
    match node with
    | :? JsonValue as jsonValue -> JsonResult.succeed jsonValue
    | Null -> JsonResult.failWithMessage "JSON node is null"
    | _ -> JsonResult.failWithMessage "JSON node is not a JSON value"

let fromStream (stream: Stream | null) =
    async {
        let! cancellationToken = Async.CancellationToken

        match stream with
        | Null -> return JsonResult.failWithMessage "Stream is null."
        | NonNull stream ->
            try
                match!
                    JsonNode.ParseAsync(stream, cancellationToken = cancellationToken)
                    |> Async.AwaitTask
                with
                | Null -> return JsonResult.failWithMessage "Stream is null or is not a valid JSON."
                | NonNull node -> return JsonResult.succeed node
            with exn ->
                return JsonError.fromException exn |> JsonResult.fail
    }

let fromBinaryData (data: BinaryData | null) =
    try
        match data with
        | Null -> JsonResult.failWithMessage "Binary data is null."
        | NonNull data ->
            match JsonNode.Parse(data) with
            | Null -> JsonResult.failWithMessage "Binary data is null or is not a valid JSON."
            | NonNull node -> JsonResult.succeed node
    with exn ->
        JsonError.fromException exn |> JsonResult.fail
